---
title: "Introduction to ChIP-Seq"
author: "Mary Piper, Radhika Khetani, Meeta Mistry (HSPH), Chris Fields (UIUC)"
date: "October 30, 2017"
output: 
  revealjs::revealjs_presentation:
    theme: solarized
    highlight: pygments
    transition: slide
    self_contained: true
    slide_level: 1
    reveal_options:
      slideNumber: true
---

# Learning Objectives

Approximate time: 30 minutes

* Understanding the experimental setup and design for ChIP-Seq experiments

---

# Introduction to ChIP-Seq

--- 

<img src="../img/chipseq_overall.png" width=400>

<small>Chromatin immunoprecipitation (ChIP) experiments isolate the chromatin from a cell and immunoprecipitate (IP) DNA fragments bound to a protein of interest. In ChIP-Seq, the DNA fragments are sequenced, enriched regions of DNA or peaks are determined, and over-represented sequence motifs and functional annotations can be identified.</small>

---

During this session we will be performing a complete workflow for ChIP-Seq analysis, starting with experimental design and generation of the raw sequencing reads and ending with functional enrichment analyses and motif discovery.

![chipseq_workflow_general](../img/chipseq_workflow_general.png)

---

# Experimental design and library preparation

---

Several steps are involved in the library preparation of protein-bound DNA fragments for sequencing:

![chipseq_experimental_workflow](../img/chipseq_experimental_workflow.png)

---

<small>

<img src="../img/chipseq_experimental_workflow.png" width="300">

> 1. After the chromatin is isolated from the cell, proteins are cross-linked to the DNA
> 2. The DNA is sheared into fragments of 200-400bp (may depend on protocol)
> 3. A protein-specific antibody is used to immunoprecipitate the protein-bound DNA fragments
> 4. Separate the antibody-bound DNA from unbound genomic DNA
> 5. The crosslink is reversed and DNA purified
> 6. DNA fragments are size selected and amplified using PCR

</small>

---

Within the DNA fragments enriched for the regions binding to a protein of interest, only a fraction correspond to actual signal. The proportion of DNA fragments containing the actual binding site of the protein depends on the **number of active binding sites, the number of starting genomes, and the efficiency of the IP**.

---

In addition, when performing ChIP-Seq, some sequences may appear enriched due to the following:

- Open chromatin regions are fragmented more easily than closed regions
- Repetitive sequences might seem to be enriched (copy number inaccuracies in genome assembly)
- Uneven distribution of sequence reads across the genome

---

Therefore, proper controls are essential. A ChIP-Seq peak should be compared with the same region of the genome in a matched control.

<img src="../img/chipseq_exp_peaks.png" width="800">

---

The same starting material should be divided to be used for both the protein-specific IP and the control. The control sample can be generated by one of the following recommended techniques:

> * No IP (input DNA)
> * No antibody ("mock IP")
> * Non-specific antibody (IgG "mock IP")

![chipseq_exp_controls](../img/chipseq_exp_controls.png)

---

# Introduction to example data

---

* Our goal for this session is to compare the the binding profiles of [Nanog](http://www.nature.com/stemcells/2009/0909/090910/full/stemcells.2009.118.html) and [Pou5f1](http://www.nature.com/cr/journal/v12/n5/full/7290134a.html) (Oct4). 
* The ChIP was performed on H1 human embryonic stem cell line (h1-ESC) cells, and sequenced using Illumina. 
* The datasets were obtained from the [HAIB TFBS ENCODE collection](http://hgdownload.cse.ucsc.edu/goldenpath/hg19/encodeDCC/wgEncodeHaibTfbs/). 
* These 2 transcription factors are involved in **stem cell pluripotency** and one of the goals is to understand their roles, individually and together, in transcriptional regulation.

---

<img src="../img/chipseq_exp_design.png" width=500>

Two replicates were collected and each was divided into 3 aliquots for the following:

- Nanog IP
- Pou5f1 IP
- Control input DNA

---

For these 6 samples, we will be using reads from only a 32.8 Mb of chromosome 12 (chr12:1,000,000-33,800,000), so we can get through the workflow in class.

Below is the workflow that we will be using today, each step in the workflow will require the data to be in a specific type of standardized format.

<img src="../img/chip_workflow_june2017.png" width="500">

---

## Set-up

---

Before we get started with the analysis, we need to set up our directory structure.

[Log in](https://help.igb.illinois.edu/Biocluster2#How_To_Log_Into_The_Cluster) to biocluster2:

```bash
# use your class ID in place of the one below
ssh class123@biologin.igb.illinois.edu

# when prompted, enter your password
```

---

Start an interactive session with two cores:

```bash
$ srun -n 2 --mem 2000 -p classroom --pty bash
```

---

Note the prompt change:

```bash
[instru03@biologin-0 ~]$ srun -c 2 --mem 2000 -p classroom --pty bash
[instru03@compute-1-0 ~]$
```

The part in brackets is your user name and the node; the node changed when you started an interactive job.

---

Create a new work directory

```bash
$ mkdir ngs_course 
$ cd ~/ngs_course
```

---

Create a `chipseq` directory and change directories into it:

```bash
$ mkdir chipseq
$ cd chipseq
```

---

Now let's setup the directory structure, we are looking for the following structure within the `chipseq` directory:

```text
chipseq/
├── logs/
├── meta/
├── raw_data/
├── reference_data/
├── results/
│   ├── bowtie2/
│   ├── trimmed/
│   ├── trimmed_fastqc/
│   └── untrimmed_fastqc/
└── scripts/
```

---

This is what we start with (using the `tree` command):

```
$ tree
.

0 directories, 0 files
```

---

Here's how we create this directory structure

```bash
$ mkdir -p raw_data reference_data scripts logs meta

$ mkdir -p results/untrimmed_fastqc results/trimmed results/trimmed_fastqc results/bowtie2
```

---

Now, type the `tree` command again:

```
$ tree
```

You should see something like this:

```
.
├── logs
├── meta
├── raw_data
├── reference_data
├── results
│   ├── bowtie2
│   ├── trimmed
│   ├── trimmed_fastqc
│   └── untrimmed_fastqc
└── scripts

10 directories, 0 files
```

---

Now that we have the directory structure created, let's copy over the data to perform our quality control and alignment, including our FASTQ files and reference data files:

```bash
$ cp /home/classroom/hpcbio/chip-seq/raw_data/*.fastq.gz raw_data/
$ cp /home/classroom/hpcbio/chip-seq/reference/* reference_data/
```

---

What does `tree` look like now?

---

```
[instru03@compute-1-0 chipseq]$ tree
.
├── logs
├── meta
├── raw_data
│   ├── H1hesc_Input_Rep1_chr12.fastq
│   ├── H1hesc_Input_Rep2_chr12.fastq
│   ├── H1hesc_Nanog_Rep1_chr12.fastq
│   ├── H1hesc_Nanog_Rep2_chr12.fastq
│   ├── H1hesc_Pou5f1_Rep1_chr12.fastq
│   └── H1hesc_Pou5f1_Rep2_chr12.fastq
├── reference_data
│   ├── chr12.1.bt2
│   ├── chr12.2.bt2
│   ├── chr12.3.bt2
│   ├── chr12.4.bt2
│   ├── chr12.fa
│   ├── chr12.rev.1.bt2
│   └── chr12.rev.2.bt2
├── results
│   ├── bowtie2
│   ├── trimmed
│   ├── trimmed_fastqc
│   └── untrimmed_fastqc
└── scripts

10 directories, 13 files
```

***
*This lesson has been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*
